File: src/genomi/interfaces/templates/portal_starter_cards.js
export function portalStarterModel() {
  const script = typeof document !== 'undefined' && document.getElementById
    ? document.getElementById('genomi-starter-cards')
    : null;
  if (!script) return { intro: '', cards: [] };
  try {
    const parsed = JSON.parse(script.textContent || '{}');
    return starterModel(parsed);
  } catch {
    return { intro: '', cards: [] };
  }
}

export function welcomeMessageMarkup(model = portalStarterModel()) {
  const data = starterModel(model);
  if (!data.intro && !data.cards.length) return '';
  return [
    '<div class="message assistant chat-welcome" data-testid="genomi-chat-welcome">',
    '<div class="role">Genomi Portal</div>',
    '<div class="body">',
    data.intro ? '<p>' + escapeHtml(data.intro) + '</p>' : '',
    '<div class="message-suggestions" data-testid="genomi-chat-suggestions">',
    data.cards.map(starterCardMarkup).join(''),
    '</div>',
    '</div>',
    '</div>'
  ].join('');
}

export function setStarterCardPending(button) {
  setStarterCardState(button, 'loading', 'Opening evidence source...');
}

export function clearStarterCardState(button) {
  if (!button) return;
  const detail = starterCardDetail(button);
  if (detail && button.dataset.starterDefaultDetail) detail.textContent = button.dataset.starterDefaultDetail;
  if (detail && detail.removeAttribute) detail.removeAttribute('data-source-unavailable');
  delete button.dataset.sourceState;
  if (button.removeAttribute) button.removeAttribute('aria-live');
}

export function markStarterSourceUnavailable(button, message = '') {
  const detail = starterCardDetail(button);
  setStarterCardState(button, 'unavailable', sourceUnavailableText(message));
  if (detail) detail.setAttribute('data-source-unavailable', 'true');
}

function setStarterCardState(button, state, message) {
  if (!button) return;
  const detail = starterCardDetail(button);
  if (detail && !button.dataset.starterDefaultDetail) {
    button.dataset.starterDefaultDetail = detail.textContent || '';
  }
  button.dataset.sourceState = state;
  button.setAttribute('aria-live', 'polite');
  if (detail) {
    detail.textContent = message;
    if (state !== 'unavailable' && detail.removeAttribute) detail.removeAttribute('data-source-unavailable');
  }
}

function sourceUnavailableText(message) {
  const detail = String(message || '').replace(/\s+/g, ' ').trim();
  return detail
    ? 'Evidence source unavailable. ' + detail
    : 'Evidence source unavailable. Refresh evidence sources or try again.';
}

function starterCardDetail(button) {
  if (!button || !button.querySelector) return null;
  return button.querySelector('[data-starter-card-detail]') || button.querySelector('span');
}

function starterModel(value) {
  const source = value && typeof value === 'object' && !Array.isArray(value) ? value : {};
  return {
    intro: String(source.intro || ''),
    cards: (Array.isArray(source.cards) ? source.cards : []).map(starterCardModel).filter(Boolean)
  };
}

function starterCardModel(value) {
  const source = value && typeof value === 'object' && !Array.isArray(value) ? value : {};
  const label = String(source.label || '').trim();
  const detail = String(source.detail || '').trim();
  const prompt = String(source.prompt || '');
  const sourceOperation = String(source.sourceOperation || '').trim();
  if (!label || !detail || (!prompt && !sourceOperation)) return null;
  const card = {
    id: String(source.id || '').trim(),
    label,
    detail,
    prompt,
    sourceOperation,
    sourceParams: object(source.sourceParams)
  };
  return card;
}

function starterCardMarkup(card) {
  const attrs = [
    ['class', 'suggestion-chip'],
    ['type', 'button'],
    ['data-starter-card-id', card.id]
  ];
  if (card.prompt) attrs.push(['data-prompt', card.prompt]);
  if (card.sourceOperation) attrs.push(['data-source-operation', card.sourceOperation]);
  let attrText = attrs
    .filter((item) => item[1])
    .map((item) => item[0] + '="' + escapeAttribute(item[1]) + '"')
    .join(' ');
  const params = sourceParamsText(card.sourceParams);
  if (params) attrText += " data-source-params='" + escapeSingleQuotedAttribute(params) + "'";
  return [
    '<button ',
    attrText,
    '>',
    '<strong>' + escapeHtml(card.label) + '</strong>',
    '<span data-starter-card-detail>' + escapeHtml(card.detail) + '</span>',
    '</button>'
  ].join('');
}

function sourceParamsText(value) {
  const params = object(value);
  if (!Object.keys(params).length) return '';
  try {
    return JSON.stringify(params);
  } catch {
    return '';
  }
}

function object(value) {
  return value && typeof value === 'object' && !Array.isArray(value) ? value : {};
}

function escapeHtml(value) {
  return String(value || '')
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;');
}

function escapeAttribute(value) {
  return escapeHtml(value)
    .replace(/"/g, '&quot;');
}

function escapeSingleQuotedAttribute(value) {
  return escapeHtml(value)
    .replace(/'/g, '&#39;');
}

File: src/genomi/interfaces/portal_starter_cards.py
from __future__ import annotations

from dataclasses import dataclass, field
import html
import json
from typing import Any, Mapping

JsonObject = dict[str, Any]

WELCOME_INTRO = (
    "Ask about symptoms, a diagnosis, a medication, or your genome. "
    "Genomi researches the question and keeps every record and source it used in this workspace."
)


@dataclass(frozen=True)
class StarterCard:
    card_id: str
    label: str
    detail: str
    prompt: str = ""
    source_operation: str = ""
    source_params: Mapping[str, Any] = field(default_factory=dict)

    def to_json(self) -> JsonObject:
        model: JsonObject = {
            "id": self.card_id,
            "label": self.label,
            "detail": self.detail,
        }
        if self.prompt:
            model["prompt"] = self.prompt
        if self.source_operation:
            model["sourceOperation"] = self.source_operation
        if self.source_params:
            model["sourceParams"] = dict(self.source_params)
        return model


STARTER_CARDS: tuple[StarterCard, ...] = (
    StarterCard(
        card_id="investigate-health-question",
        label="Investigate a health question",
        detail=(
            "Bring symptoms, diagnoses, and test results. Genomi keeps competing explanations, "
            "evidence, and open questions together as the investigation grows."
        ),
        prompt=(
            "I have been dealing with the following: describe your symptoms, when they started, "
            "and anything that makes them better or worse. My diagnoses so far: list them. "
            "Medications I take: list them. What I want to understand: say it in your own words."
        ),
    ),
    StarterCard(
        card_id="review-a-medication",
        label="Review a medication",
        detail="Ask how your genetics could change how a drug you take, or are about to start, works for you.",
        prompt=(
            "How could my genetics affect the way I respond to this medication: name the drug. "
            "I am taking it for: say the reason."
        ),
    ),
    StarterCard(
        card_id="add-genome",
        label="Add your genome",
        detail="Point Genomi at a 23andMe, AncestryDNA, MyHeritage, VCF, or BAM export saved on this computer.",
        source_operation="genomi.parse_source",
    ),
)


def welcome_model() -> JsonObject:
    return {
        "intro": WELCOME_INTRO,
        "cards": [card.to_json() for card in STARTER_CARDS],
    }


def welcome_model_json() -> str:
    return json.dumps(welcome_model(), separators=(",", ":"), sort_keys=True).replace("</", "<\\/")


def welcome_message_markup() -> str:
    return "".join(
        (
            '<div class="message assistant">',
            '<div class="role">GenomiLab</div>',
            '<div class="body">',
            f"<p>{html.escape(WELCOME_INTRO)}</p>",
            '<div class="message-suggestions" data-testid="genomi-chat-suggestions">',
            "".join(_card_markup(card) for card in STARTER_CARDS),
            "</div>",
            "</div>",
            "</div>",
        )
    )


def _card_markup(card: StarterCard) -> str:
    attrs = [
        ('class', 'suggestion-chip'),
        ('type', 'button'),
        ('data-starter-card-id', card.card_id),
    ]
    if card.prompt:
        attrs.append(('data-prompt', card.prompt))
    if card.source_operation:
        attrs.append(('data-source-operation', card.source_operation))
    attr_text = " ".join(_double_quoted_attr(name, value) for name, value in attrs)
    if card.source_params:
        params = json.dumps(dict(card.source_params), separators=(",", ":"), sort_keys=True)
        attr_text += " " + _single_quoted_attr("data-source-params", params)
    return (
        f"<button {attr_text}>"
        f"<strong>{html.escape(card.label)}</strong>"
        f'<span data-starter-card-detail>{html.escape(card.detail)}</span>'
        "</button>"
    )


def _double_quoted_attr(name: str, value: str) -> str:
    return f'{name}="{html.escape(value, quote=True)}"'


def _single_quoted_attr(name: str, value: str) -> str:
    escaped = html.escape(value, quote=False).replace("'", "&#x27;")
    return f"{name}='{escaped}'"
